test/library_search/basic.lean

```diff
@@ -0,0 +1,224 @@
+
+-- Enable this option for tracing:
+-- set_option trace.Tactic.stdLibrarySearch true
+-- And this option to trace all candidate lemmas before application.
+-- set_option trace.Tactic.stdLibrarySearch.lemmas true
+
+noncomputable section
+
+/-- info: Try this: exact Nat.lt.base x -/
+#guard_msgs in
+example (x : Nat) : x ≠ x.succ := Nat.ne_of_lt (by std_apply?)
+
+/-- info: Try this: exact Nat.zero_lt_succ 1 -/
+#guard_msgs in
+example : 0 ≠ 1 + 1 := Nat.ne_of_lt (by std_apply?)
+
+example : 0 ≠ 1 + 1 := Nat.ne_of_lt (by exact Fin.size_pos')
+
+/-- info: Try this: exact Nat.add_comm x y -/
+#guard_msgs in
+example (x y : Nat) : x + y = y + x := by std_apply?
+
+/-- info: Try this: exact fun a => Nat.add_le_add_right a k -/
+#guard_msgs in
+example (n m k : Nat) : n ≤ m → n + k ≤ m + k := by std_apply?
+
+/-- info: Try this: exact Nat.mul_dvd_mul_left a w -/
+#guard_msgs in
+example (ha : a > 0) (w : b ∣ c) : a * b ∣ a * c := by std_apply?
+
+-- Could be any number of results (`Int.one`, `Int.zero`, etc)
+#guard_msgs (drop info) in
+example : Int := by std_apply?
+
+/-- info: Try this: Nat.lt.base x -/
+#guard_msgs in
+example : x < x + 1 := std_exact?%
+
+/-- info: Try this: exact p -/
+#guard_msgs in
+example (P : Prop) (p : P) : P := by std_apply?
+/-- info: Try this: exact False.elim (np p) -/
+#guard_msgs in
+example (P : Prop) (p : P) (np : ¬P) : false := by std_apply?
+/-- info: Try this: exact h x rfl -/
+#guard_msgs in
+example (X : Type) (P : Prop) (x : X) (h : ∀ x : X, x = x → P) : P := by std_apply?
+
+-- Could be any number of results (`fun x => x`, `id`, etc)
+#guard_msgs (drop info) in
+example (α : Prop) : α → α := by std_apply?
+
+-- Note: these examples no longer work after we turned off lemmas with discrimination key `#[*]`.
+-- example (p : Prop) : (¬¬p) → p := by std_apply? -- says: `exact not_not.mp`
+-- example (a b : Prop) (h : a ∧ b) : a := by std_apply? -- says: `exact h.left`
+-- example (P Q : Prop) : (¬ Q → ¬ P) → (P → Q) := by std_apply? -- say: `exact Function.mtr`
+
+/-- info: Try this: exact Nat.add_comm a b -/
+#guard_msgs in
+example (a b : Nat) : a + b = b + a :=
+by std_apply?
+
+/-- info: Try this: exact Nat.mul_sub_left_distrib n m k -/
+#guard_msgs in
+example (n m k : Nat) : n * (m - k) = n * m - n * k :=
+by std_apply?
+
+attribute [symm] Eq.symm
+
+/-- info: Try this: exact Eq.symm (Nat.mul_sub_left_distrib n m k) -/
+#guard_msgs in
+example (n m k : Nat) : n * m - n * k = n * (m - k) := by
+  std_apply?
+
+/-- info: Try this: exact eq_comm -/
+#guard_msgs in
+example {α : Type} (x y : α) : x = y ↔ y = x := by std_apply?
+
+/-- info: Try this: exact Nat.add_pos_left ha b -/
+#guard_msgs in
+example (a b : Nat) (ha : 0 < a) (_hb : 0 < b) : 0 < a + b := by std_apply?
+
+/-- info: Try this: exact Nat.add_pos_left ha b -/
+#guard_msgs in
+-- Verify that if maxHeartbeats is 0 we don't stop immediately.
+set_option maxHeartbeats 0 in
+example (a b : Nat) (ha : 0 < a) (_hb : 0 < b) : 0 < a + b := by std_apply?
```

Comment:
While I remember that deterministic testing of the maxHeartbeats feature was hard to arrange in the original version, I presume that you've verified observationally at least once that the new implementation of this feature works. :-)

Reply:
I did a quick test.  I am going to leave this as is, but I think it'd be good to clarify the purpose of this, and see if there is a better way to achieve it.  My impression is the feature is so users get back some results.